File: Realize/3Week/PracticeQueue.cpp
#include <iostream>
using namespace std;

class MyQueue2 {
private:
    int* arr;
    int front;
    int rear;
    int size;

public:
    MyQueue2(int size) {
        this->size = size + 1;
        arr = new int[this->size];
        front = 0;
        rear = 0;
    }

    ~MyQueue2() {
        delete[] arr;
    }

    bool isEmpty() {
        return rear == front;
    }

    bool isFull() {
        return (rear + 1) % size == front;
    }

    void enqueue(int data) {
        if (isFull()) {
            cout << "Queue is full!" << endl;
            return;
        }
        rear = (rear + 1) % size;
        arr[rear] = data;
    }

    int dequeue() {
        if (isEmpty()) {
            cout << "Queue is empty!" << endl;
            return -1;
        }
        front = (front + 1) % size;
        return arr[front];
    }

    void printQueue() {
        int start = (front + 1) % size;
        int end = (rear + 1) % size;
        for (int i = start; i != end; i = (i + 1) % size) {
            cout << arr[i] << " ";
        }
        cout << endl;
    }
};

int main() {
    MyQueue2 myQueue(5);
    myQueue.enqueue(1);
    myQueue.enqueue(2);
    myQueue.enqueue(3);
    myQueue.enqueue(4);
    myQueue.enqueue(5);
    myQueue.enqueue(6); // Queue is full!

    myQueue.printQueue(); // 1 2 3 4 5

    cout << myQueue.dequeue() << endl; // 1
    myQueue.printQueue(); // 2 3 4 5

    cout << myQueue.dequeue() << endl; // 2
    myQueue.printQueue(); // 3 4 5

    myQueue.enqueue(6);
    myQueue.enqueue(7);
    myQueue.printQueue(); // 3 4 5 6 7

    cout << myQueue.dequeue() << endl; // 3
    cout << myQueue.dequeue() << endl; // 4
    cout << myQueue.dequeue() << endl; // 5
    myQueue.printQueue(); // 6 7
    cout << myQueue.dequeue() << endl; // 6
    cout << myQueue.dequeue() << endl; // 7
    myQueue.dequeue(); // Queue is empty!

    return 0;
}

File: Realize/3Week/PracticeQueue.js
class MyQueue2 {
    constructor(size) {
        this.size = size + 1;
        this.arr = new Array(this.size).fill(0);
        this.front = 0;
        this.rear = 0;
    }

    isEmpty() {
        return this.rear === this.front;
    }

    isFull() {
        return (this.rear + 1) % this.size === this.front;
    }

    enqueue(data) {
        if (this.isFull()) {
            console.log("Queue is full!");
            return;
        }
        this.rear = (this.rear + 1) % this.size;
        this.arr[this.rear] = data;
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log("Queue is empty!");
            return null;
        }
        this.front = (this.front + 1) % this.size;
        return this.arr[this.front];
    }

    printQueue() {
        let start = (this.front + 1) % this.size;
        let end = (this.rear + 1) % this.size;
        let i = start;
        let result = "";
        while (i !== end) {
            result += this.arr[i] + " ";
            i = (i + 1) % this.size;
        }
        console.log(result.trim());
    }
}

const myQueue = new MyQueue2(5);
myQueue.enqueue(1);
myQueue.enqueue(2);
myQueue.enqueue(3);
myQueue.enqueue(4);
myQueue.enqueue(5);
myQueue.enqueue(6); // Queue is full!

myQueue.printQueue(); // 1 2 3 4 5

console.log(myQueue.dequeue()); // 1
myQueue.printQueue(); // 2 3 4 5

console.log(myQueue.dequeue()); // 2
myQueue.printQueue(); // 3 4 5

myQueue.enqueue(6);
myQueue.enqueue(7);
myQueue.printQueue(); // 3 4 5 6 7

console.log(myQueue.dequeue()); // 3
console.log(myQueue.dequeue()); // 4
console.log(myQueue.dequeue()); // 5
myQueue.printQueue(); // 6 7
console.log(myQueue.dequeue()); // 6
console.log(myQueue.dequeue()); // 7
myQueue.dequeue(); // Queue is empty!
